crates/solidity/inputs/language/src/definition.rs

```diff
@@ -3692,6 +3692,21 @@ codegen_language_macros::compile!(Language(
                                         not_followed_by = Fragment(IdentifierStart)
                                     )
                                 ),
+                                // Since 0.5.0, only consume a dot if it's followed by a digit:
```

Comment:
Not sure I understand the fix here:

- For input `5`, it will always be matched by the first definition.
- For input `5.`, it will be matched by the second definition till version `0.5.0`, and rejected afterwards.
- For input `5.6`, it will always be matched by the last definition.

Inserting this new definition into the 3rd position just overrides the last one, right?

Reply:
I see. Thanks for explaining!
I think my concern here is that the new definition basically supersedes other definitions. In this case, what do you think of the following:

- `An integer (without a dot or a fraction) is enabled in all versions`
    - stays the same
- `An integer and a dot (without a fraction) is enabled till "0.5.0"`
    - add the `fraction` here and make it optional
- `An integer, a dot, and a fraction is enabled from "0.5.0"`
    - add this new one (all three are required)
- `A dot and a fraction (without an integer) is enabled in all versions`
    - stays the same
- `An integer, a dot, and a fraction is enabled in all versions`
    - removed, as it is replaced by the second definition.

This has the additional benefit of keeping the specific scanners mutually exclusive, which is going to allow building a state machine for the scanner. Most of the `TokenDefinition`s in the DSL are built this way.

Reply:
That sounds good in theory and I've tried it, but we still need to be able to optionally skip the `(. DECIMAL_DIGITS)`. That is because the first rule has to have a negative lookahead for `.`; otherwise we end up accepting `5.member()` as valid in < 0.5.0.

I've pushed a version that is hopefully more acceptable, since a bit more disjoint now. Whenever we move to compiling these down to a state machine, I'm happy to revisit the definitions 🤞